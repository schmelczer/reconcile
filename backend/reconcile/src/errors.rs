use thiserror::Error;

#[derive(Error, Debug)]
pub enum SyncLibError {
    #[error("Failed to apply operation because {0}")]
    OperationApplicationError(String),
}
