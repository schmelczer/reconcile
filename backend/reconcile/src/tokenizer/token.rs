#[derive(Debug, Clone)]
pub struct Token<T>
where
    T: PartialEq + Clone,
{
    normalised: T,
    original: String,
}

impl From<&str> for Token<String> {
    fn from(s: &str) -> Self {
        Token {
            normalised: s.to_string(),
            original: s.to_string(),
        }
    }
}

impl<T> Token<T>
where
    T: PartialEq + Clone,
{
    pub fn new(normalised: T, original: String) -> Self {
        Token {
            normalised,
            original,
        }
    }

    pub fn original(&self) -> &str {
        &self.original
    }

    pub fn normalised(&self) -> &T {
        &self.normalised
    }

    pub fn get_original_length(&self) -> usize {
        self.original.chars().count()
    }
}

impl<T> PartialEq for Token<T>
where
    T: PartialEq + Clone,
{
    fn eq(&self, other: &Self) -> bool {
        self.normalised == other.normalised
    }
}
