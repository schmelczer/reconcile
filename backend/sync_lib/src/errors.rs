use base64::DecodeError;
use thiserror::Error;
use wasm_bindgen::JsValue;

#[derive(Error, Debug)]
pub enum SyncLibError {
    #[error("Base64 decoding error: {}", .reason)]
    DecodingError { reason: String },
}

impl From<DecodeError> for SyncLibError {
    fn from(e: DecodeError) -> Self {
        SyncLibError::DecodingError {
            reason: e.to_string(),
        }
    }
}

impl From<std::string::FromUtf8Error> for SyncLibError {
    fn from(e: std::string::FromUtf8Error) -> Self {
        SyncLibError::DecodingError {
            reason: e.to_string(),
        }
    }
}

impl From<SyncLibError> for JsValue {
    fn from(val: SyncLibError) -> Self { JsValue::from_str(&val.to_string()) }
}
