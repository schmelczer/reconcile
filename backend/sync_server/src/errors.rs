use std::fmt::Display;

use axum::{
    Json,
    http::StatusCode,
    response::{IntoResponse, Response},
};
use log::{debug, error};
use serde::Serialize;
use thiserror::Error;
use ts_rs::TS;

#[derive(Error, Debug)]
pub enum SyncServerError {
    #[error("Initialisation error: {0}")]
    InitError(#[source] anyhow::Error),

    #[error("Client error: {0:?}")]
    ClientError(#[source] anyhow::Error),

    #[error("Server error: {0:?}")]
    ServerError(#[source] anyhow::Error),

    #[error("Not found: {0}")]
    NotFound(#[source] anyhow::Error),

    #[error("Unauthorized: {0}")]
    Unauthenticated(#[source] anyhow::Error),

    #[error("Permission denied error: {0}")]
    PermissionDeniedError(#[source] anyhow::Error),
}

impl SyncServerError {
    pub fn serialize(&self) -> SerializedError {
        match self {
            Self::InitError(error)
            | Self::ClientError(error)
            | Self::ServerError(error)
            | Self::NotFound(error)
            | Self::Unauthenticated(error)
            | Self::PermissionDeniedError(error) => error.into(),
        }
    }
}

#[derive(TS, Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct SerializedError {
    pub error_type: &'static str,
    pub message: String,
    pub causes: Vec<String>,
}

impl Display for SerializedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if !self.causes.is_empty() {
            write!(f, "\nCauses:\n")?;
            for cause in &self.causes {
                write!(f, "{}", &format!("- {cause}\n"))?;
            }
        }

        Ok(())
    }
}

impl IntoResponse for SyncServerError {
    fn into_response(self) -> Response {
        let body = Json(self.serialize());

        match self {
            Self::InitError(_) | Self::ServerError(_) => {
                (StatusCode::INTERNAL_SERVER_ERROR, body).into_response()
            }
            Self::ClientError(_) => (StatusCode::BAD_REQUEST, body).into_response(),
            Self::NotFound(_) => (StatusCode::NOT_FOUND, body).into_response(),
            Self::Unauthenticated(_) => (StatusCode::UNAUTHORIZED, body).into_response(),
            Self::PermissionDeniedError(_) => (StatusCode::FORBIDDEN, body).into_response(),
        }
    }
}

impl From<&anyhow::Error> for SerializedError {
    fn from(error: &anyhow::Error) -> SerializedError {
        let mut causes = vec![];
        let mut current_error = error.source();
        while let Some(error) = current_error {
            causes.push(error.to_string());
            current_error = error.source();
        }

        SerializedError {
            error_type: error.downcast_ref::<SyncServerError>().map_or(
                "UnknownError",
                |e| match e {
                    SyncServerError::InitError(_) => "InitError",
                    SyncServerError::ClientError(_) => "ClientError",
                    SyncServerError::ServerError(_) => "ServerError",
                    SyncServerError::NotFound(_) => "NotFound",
                    SyncServerError::Unauthenticated(_) => "Unauthenticated",
                    SyncServerError::PermissionDeniedError(_) => "PermissionDeniedError",
                },
            ),
            message: error.to_string(),
            causes,
        }
    }
}

pub fn init_error(error: anyhow::Error) -> SyncServerError {
    debug!("Initialization error: {error:?}");
    SyncServerError::InitError(error)
}

pub fn server_error(error: anyhow::Error) -> SyncServerError {
    debug!("Server error: {error:?}");
    SyncServerError::ServerError(error)
}

pub fn client_error(error: anyhow::Error) -> SyncServerError {
    debug!("Client error: {error:?}");
    SyncServerError::ClientError(error)
}

pub fn not_found_error(error: anyhow::Error) -> SyncServerError {
    debug!("Not found: {error:?}");
    SyncServerError::NotFound(error)
}

pub fn unauthenticated_error(error: anyhow::Error) -> SyncServerError {
    debug!("Unauthenticated user: {error:?}");
    SyncServerError::Unauthenticated(error)
}

pub fn permission_denied_error(error: anyhow::Error) -> SyncServerError {
    debug!("Permission denied: {error:?}");
    SyncServerError::PermissionDeniedError(error)
}
