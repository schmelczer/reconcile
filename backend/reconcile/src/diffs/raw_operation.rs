use crate::tokenizer::token::Token;

#[derive(Debug, Clone, PartialEq)]
pub enum RawOperation {
    Insert(Vec<Token>),
    Delete(Vec<Token>),
    Equal(Vec<Token>),
}

impl RawOperation {
    pub fn tokens(&self) -> &Vec<Token> {
        match self {
            RawOperation::Insert(tokens) => tokens,
            RawOperation::Delete(tokens) => tokens,
            RawOperation::Equal(tokens) => tokens,
        }
    }

    pub fn original_text_length(&self) -> usize {
        self.tokens()
            .iter()
            .map(|t| t.original.chars().count())
            .sum()
    }

    pub fn get_original_text(self) -> String {
        self.tokens().iter().map(|t| t.original.clone()).collect()
    }
}
