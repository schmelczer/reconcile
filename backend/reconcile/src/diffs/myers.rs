//! Taken from https://github.com/mitsuhiko/similar/blob/7e15c44de11a1cd61e1149189929e189ef977fd8/src/algorithms/myers.rs
//! Myers' diff algorithm.
//!
//! * time: `O((N+M)D)`
//! * space `O(N+M)`
//!
//! See [the original article by Eugene W. Myers](http://www.xmailserver.org/diff2.pdf)
//! describing it.
//!
//! The implementation of this algorithm is based on the implementation by
//! Brandon Williams.
//!
//! # Heuristics
//!
//! At present this implementation of Myers' does not implement any more
//! advanced heuristics that would solve some pathological cases.  For instance
//! passing two large and completely distinct sequences to the algorithm will
//! make it spin without making reasonable progress.
//! For potential improvements here see [similar#15](https://github.com/mitsuhiko/similar/issues/15).

use std::{
    ops::{Index, IndexMut, Range},
    vec,
};

use super::raw_operation::RawOperation;
use crate::{
    tokenizer::token::Token,
    utils::{common_prefix_len::common_prefix_len, common_suffix_len::common_suffix_len},
};

/// Myers' diff algorithm with deadline.
///
/// Diff `old`, between indices `old_range` and `new` between indices
/// `new_range`.
///
/// This diff is done with an optional deadline that defines the maximal
/// execution time permitted before it bails and falls back to an approximation.
pub fn diff<T>(old: &[Token<T>], new: &[Token<T>]) -> Vec<RawOperation<T>>
where
    T: PartialEq + Clone,
{
    let max_d = max_d(old.len(), new.len());
    let mut vb = V::new(max_d);
    let mut vf = V::new(max_d);
    let mut result: Vec<RawOperation<T>> = vec![];
    conquer(
        old,
        0..old.len(),
        new,
        0..new.len(),
        &mut vf,
        &mut vb,
        &mut result,
    );
    result
}

// A D-path is a path which starts at (0,0) that has exactly D non-diagonal
// edges. All D-paths consist of a (D - 1)-path followed by a non-diagonal edge
// and then a possibly empty sequence of diagonal edges called a snake.

/// `V` contains the endpoints of the furthest reaching `D-paths`. For each
/// recorded endpoint `(x,y)` in diagonal `k`, we only need to retain `x`
/// because `y` can be computed from `x - k`. In other words, `V` is an array of
/// integers where `V[k]` contains the row index of the endpoint of the furthest
/// reaching path in diagonal `k`.
///
/// We can't use a traditional Vec to represent `V` since we use `k` as an index
/// and it can take on negative values. So instead `V` is represented as a
/// light-weight wrapper around a Vec plus an `offset` which is the maximum
/// value `k` can take on in order to map negative `k`'s back to a value >= 0.
#[derive(Debug)]
struct V {
    offset: isize,
    v: Vec<usize>, // Look into initializing this to -1 and storing isize
}

impl V {
    fn new(max_d: usize) -> Self {
        Self {
            offset: max_d as isize,
            v: vec![0; 2 * max_d],
        }
    }

    fn len(&self) -> usize { self.v.len() }
}

impl Index<isize> for V {
    type Output = usize;

    fn index(&self, index: isize) -> &Self::Output { &self.v[(index + self.offset) as usize] }
}

impl IndexMut<isize> for V {
    fn index_mut(&mut self, index: isize) -> &mut Self::Output {
        &mut self.v[(index + self.offset) as usize]
    }
}

fn max_d(len1: usize, len2: usize) -> usize {
    // XXX look into reducing the need to have the additional '+ 1'
    (len1 + len2 + 1) / 2 + 1
}

#[inline(always)]
fn split_at(range: Range<usize>, at: usize) -> (Range<usize>, Range<usize>) {
    (range.start..at, at..range.end)
}

/// A `Snake` is a sequence of diagonal edges in the edit graph.  Normally
/// a snake has a start end end point (and it is possible for a snake to have
/// a length of zero, meaning the start and end points are the same) however
/// we do not need the end point which is why it's not implemented here.
///
/// The divide part of a divide-and-conquer strategy. A D-path has D+1 snakes
/// some of which may be empty. The divide step requires finding the ceil(D/2) +
/// 1 or middle snake of an optimal D-path. The idea for doing so is to
/// simultaneously run the basic algorithm in both the forward and reverse
/// directions until furthest reaching forward and reverse paths starting at
/// opposing corners 'overlap'.
fn find_middle_snake<T>(
    old: &[Token<T>],
    old_range: Range<usize>,
    new: &[Token<T>],
    new_range: Range<usize>,
    vf: &mut V,
    vb: &mut V,
) -> Option<(usize, usize)>
where
    T: PartialEq + Clone,
{
    let n = old_range.len();
    let m = new_range.len();

    // By Lemma 1 in the paper, the optimal edit script length is odd or even as
    // `delta` is odd or even.
    let delta = n as isize - m as isize;
    let odd = delta & 1 == 1;

    // The initial point at (0, -1)
    vf[1] = 0;
    // The initial point at (N, M+1)
    vb[1] = 0;

    // We only need to explore ceil(D/2) + 1
    let d_max = max_d(n, m);
    assert!(vf.len() >= d_max);
    assert!(vb.len() >= d_max);

    for d in 0..d_max as isize {
        // Forward path
        for k in (-d..=d).rev().step_by(2) {
            let mut x = if k == -d || (k != d && vf[k - 1] < vf[k + 1]) {
                vf[k + 1]
            } else {
                vf[k - 1] + 1
            };
            let y = (x as isize - k) as usize;

            // The coordinate of the start of a snake
            let (x0, y0) = (x, y);
            //  While these sequences are identical, keep moving through the
            //  graph with no cost
            if x < old_range.len() && y < new_range.len() {
                let advance = common_prefix_len(
                    old,
                    old_range.start + x..old_range.end,
                    new,
                    new_range.start + y..new_range.end,
                );
                x += advance;
            }

            // This is the new best x value
            vf[k] = x;

            // Only check for connections from the forward search when N - M is
            // odd and when there is a reciprocal k line coming from the other
            // direction.
            if odd && (k - delta).abs() <= (d - 1) {
                // TODO optimize this so we don't have to compare against n
                if vf[k] + vb[-(k - delta)] >= n {
                    // Return the snake
                    return Some((x0 + old_range.start, y0 + new_range.start));
                }
            }
        }

        // Backward path
        for k in (-d..=d).rev().step_by(2) {
            let mut x = if k == -d || (k != d && vb[k - 1] < vb[k + 1]) {
                vb[k + 1]
            } else {
                vb[k - 1] + 1
            };
            let mut y = (x as isize - k) as usize;

            // The coordinate of the start of a snake
            if x < n && y < m {
                let advance = common_suffix_len(
                    old,
                    old_range.start..old_range.start + n - x,
                    new,
                    new_range.start..new_range.start + m - y,
                );
                x += advance;
                y += advance;
            }

            // This is the new best x value
            vb[k] = x;

            if !odd && (k - delta).abs() <= d {
                // TODO optimize this so we don't have to compare against n
                if vb[k] + vf[-(k - delta)] >= n {
                    // Return the snake
                    return Some((n - x + old_range.start, m - y + new_range.start));
                }
            }
        }

        // TODO: Maybe there's an opportunity to optimize and bail early?
    }

    None
}

fn conquer<T>(
    old: &[Token<T>],
    mut old_range: Range<usize>,
    new: &[Token<T>],
    mut new_range: Range<usize>,
    vf: &mut V,
    vb: &mut V,
    result: &mut Vec<RawOperation<T>>,
) where
    T: PartialEq + Clone,
{
    // Check for common prefix
    let common_prefix_len = common_prefix_len(old, old_range.clone(), new, new_range.clone());
    if common_prefix_len > 0 {
        result.push(RawOperation::Equal(
            old[old_range.start..old_range.start + common_prefix_len].to_vec(),
        ));
    }
    old_range.start += common_prefix_len;
    new_range.start += common_prefix_len;

    // Check for common suffix
    let common_suffix_len = common_suffix_len(old, old_range.clone(), new, new_range.clone());
    let common_suffix = (
        old_range.end - common_suffix_len,
        new_range.end - common_suffix_len,
    );
    old_range.end -= common_suffix_len;
    new_range.end -= common_suffix_len;

    if old_range.is_empty() && new_range.is_empty() {
        // Do nothing
    } else if new_range.is_empty() {
        result.push(RawOperation::Delete(
            old[old_range.start..old_range.start + old_range.len()].to_vec(),
        ));
    } else if old_range.is_empty() {
        result.push(RawOperation::Insert(
            new[new_range.start..new_range.start + new_range.len()].to_vec(),
        ));
    } else if let Some((x_start, y_start)) =
        find_middle_snake(old, old_range.clone(), new, new_range.clone(), vf, vb)
    {
        let (old_a, old_b) = split_at(old_range, x_start);
        let (new_a, new_b) = split_at(new_range, y_start);
        conquer(old, old_a, new, new_a, vf, vb, result);
        conquer(old, old_b, new, new_b, vf, vb, result);
    } else {
        result.push(RawOperation::Delete(
            old[old_range.start..old_range.end].to_vec(),
        ));
        result.push(RawOperation::Insert(
            new[new_range.start..new_range.end].to_vec(),
        ));
    }

    if common_suffix_len > 0 {
        result.push(RawOperation::Equal(
            old[common_suffix.0..common_suffix.0 + common_suffix_len].to_vec(),
        ));
    }
}
