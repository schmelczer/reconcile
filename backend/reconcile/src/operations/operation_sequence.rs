use std::{cmp::Ordering, result, vec};

use super::Operation;
use crate::diffs::myers::diff;
use crate::diffs::raw_operation::RawOperation;
use crate::errors::SyncLibError;
use crate::tokenizer::token::Token;
use itertools::Itertools;
use ropey::Rope;

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Default)]
struct MergeContext {
    last_delete: Option<Operation>,
    shift: i64,
}

/// A sequence of operations that can be applied to a text document.
/// OperationSequence supports merging two sequences of operations using the
/// principle of Operational Transformation.
///
/// It's mainly created through the from_strings method, then merged with another
/// OperationSequence derived from the same original text and then applied to the original text
/// to get the reconciled text of concurrent edits.
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
pub struct OperationSequence {
    operations: Vec<(usize, Operation)>,
}

impl OperationSequence {
    /// Creates a new OperationSequence with the given operations.
    /// The operations should be in the order they should be applied.
    /// The operations must not overlap.
    pub fn new(operations: Vec<(usize, Operation)>) -> Self {
        operations.iter().zip(operations.iter().skip(1)).for_each(
            |((i_prev, previous), (i_next, next))| {
                debug_assert!(
                    i_prev == i_next
                        || i_prev + previous.len() <= *i_next
                        || !(matches!(previous, Operation::Delete { .. })
                            && matches!(next, Operation::Insert { .. })),
                    "{} and {} overlap with old index {i_prev} and {i_next}",
                    previous,
                    next
                );
                debug_assert!(
                    previous.start_index() <= next.start_index(),
                    "{} must not come before {} yet it does",
                    previous,
                    next
                );
            },
        );

        Self {
            operations, // operations: Self::merge_subsequent_operations(operations),
        }
    }

    /// Creates an OperationSequence from the given original (old) and updated (new) strings.
    /// The returned OperationSequence represents the changes from the original to the updated text.
    /// When the return value is applied to the original text, it will result in the updated text.
    pub fn from_strings(original: &str, updated: &str) -> Self {
        let original_tokens = Token::tokenize(original);
        let updated_tokens = Token::tokenize(updated);

        let diff: Vec<RawOperation> = diff(&original_tokens, &updated_tokens);

        Self::new(Self::cook_operations(diff))
    }

    fn cook_operations(raw_operations: Vec<RawOperation>) -> Vec<(usize, Operation)> {
        let mut new_index = 0;
        let mut old_index = 0;

        raw_operations
            .into_iter()
            .flat_map(|raw_operation| {
                let length = raw_operation.original_text_length();

                let operation = match raw_operation {
                    RawOperation::Equal(..) => {
                        new_index += length;
                        old_index += length;

                        None
                    }
                    RawOperation::Insert(..) => {
                        let op =
                            Operation::create_insert(new_index, raw_operation.get_original_text())
                                .map(|op| (old_index, op));

                        new_index += length;

                        op
                    }
                    RawOperation::Delete(..) => {
                        let op = Operation::create_delete_with_text(
                            new_index,
                            raw_operation.get_original_text(),
                        )
                        .map(|op| (old_index, op));

                        old_index += length;

                        op
                    }
                };

                operation.into_iter()
            })
            .sorted_by_key(|(order, _)| *order)
            .collect()
    }

    pub fn merge(&self, other: &Self) -> Result<Self, SyncLibError> {
        let mut merged_operations: Vec<Operation> =
            Vec::with_capacity(self.operations.len() + other.operations.len());

        let mut left_merge_context = MergeContext::default();
        let mut right_merge_context = MergeContext::default();

        let mut left_index: usize = 0;
        let mut right_index: usize = 0;

        loop {
            let left_op = self.operations.get(left_index);
            let right_op = other.operations.get(right_index);

            let order = left_op
                .map(|(order, _)| order)
                .cmp(&right_op.map(|(order, _)| order));

            println!("left_op: {:#?} <> right_op: {:#?}", left_op, right_op);

            let left_op = left_op.map(|(_, op)| op);
            let right_op = right_op.map(|(_, op)| op);

            // let order = if order == Ordering::Equal {
            //     match (left_op.as_ref(), right_op.as_ref()) {
            //         (Some(Operation::Insert { .. }), Some(Operation::Delete { .. })) => {
            //             Ordering::Greater
            //         }
            //         (Some(Operation::Delete { .. }), Some(Operation::Insert { .. })) => {
            //             Ordering::Less
            //         }
            //         _ => Ordering::Equal,
            //     }
            // } else {
            //     order
            // };

            // debug_assert!(
            //     right_merge_context.last_delete.is_none()
            //         || left_merge_context.last_delete.is_none(),
            //     "Both contexts have a last delete"
            // );

            match (left_op, right_op, order) {
                (Some(left_op), None, _)
                | (Some(left_op), Some(_), std::cmp::Ordering::Less | std::cmp::Ordering::Equal) => {
                    Self::pick_up_dangling_delete_from_affecting_context(
                        left_op.start_index(),
                        &mut right_merge_context,
                    );

                    if let Some(op) = Self::merge_operations_with_context(
                        left_op.with_shifted_index(right_merge_context.shift)?,
                        &mut right_merge_context,
                        &mut left_merge_context,
                    )? {
                        // println!("merged {:#?}", &op);
                        if let Some(last) = merged_operations.last() {
                            debug_assert!(op.start_index() >= last.start_index());
                        }
                        merged_operations.push(op);
                    }

                    left_index += 1;
                }
                (None, Some(right_op), _)
                | (Some(_), Some(right_op), std::cmp::Ordering::Greater) => {
                    Self::pick_up_dangling_delete_from_affecting_context(
                        right_op.start_index(),
                        &mut left_merge_context,
                    );

                    if let Some(op) = Self::merge_operations_with_context(
                        right_op.with_shifted_index(left_merge_context.shift)?,
                        &mut left_merge_context,
                        &mut right_merge_context,
                    )? {
                        // println!("merged {:#?}", &op);
                        if let Some(last) = merged_operations.last() {
                            debug_assert!(op.start_index() >= last.start_index());
                        }
                        merged_operations.push(op);
                    }

                    right_index += 1;
                }
                (None, None, _) => {
                    break;
                }
            };

            println!(
                "{:#?} <> {:#?}\n\n\n",
                left_merge_context, right_merge_context
            );
        }

        println!("merged_operations: {:#?}", merged_operations.to_vec());

        Ok(Self::new(
            merged_operations.into_iter().map(|op| (0, op)).collect(),
        ))
    }

    pub fn apply<'a>(&self, rope_text: &'a mut Rope) -> Result<&'a mut Rope, SyncLibError> {
        for (_, operation) in &self.operations {
            operation.apply(rope_text)?;
        }

        Ok(rope_text)
    }

    fn merge_operations_with_context(
        aligned_operation: Operation,
        affecting_context: &mut MergeContext,
        produced_context: &mut MergeContext,
    ) -> Result<Option<Operation>, SyncLibError> {
        Ok(
            match (aligned_operation, affecting_context.last_delete.clone()) {
                (operation @ Operation::Insert { .. }, None) => {
                    produced_context.shift += operation.len() as i64;
                    Some(operation)
                }

                (operation @ Operation::Delete { .. }, None) => {
                    Self::replace_delete_in_produced_context(
                        produced_context,
                        Some(operation.clone()),
                    );
                    Some(operation)
                }

                (operation @ Operation::Insert { .. }, Some(last_delete)) => {
                    produced_context.shift += operation.len() as i64;

                    debug_assert!(
                        last_delete.range().contains(&operation.start_index()),
                        "There is a last delete ({last_delete}) but the operation ({operation}) is not contained in it"
                    );

                    let difference =
                        operation.start_index() as i64 - last_delete.start_index() as i64;

                    let moved_operation = operation.with_index(last_delete.start_index());

                    affecting_context.last_delete = Operation::create_delete(
                        moved_operation.end_index() + 1,
                        (last_delete.len() as i64 - difference) as usize,
                    );
                    affecting_context.shift -= difference;

                    Some(moved_operation)
                }

                (operation @ Operation::Delete { .. }, Some(last_delete)) => {
                    debug_assert!(
                        last_delete.range().contains(&operation.start_index()),
                        "There is a last delete ({last_delete}) but the operation ({operation}) is not contained in it"
                    );

                    let difference =
                        operation.start_index() as i64 - last_delete.start_index() as i64;

                    let updated_delete = Operation::create_delete(
                        last_delete.start_index(),
                        0.max(operation.end_index() as i64 - last_delete.end_index() as i64)
                            as usize,
                    );

                    affecting_context.shift -= difference;
                    affecting_context.last_delete = Operation::create_delete(
                        last_delete.start_index(),
                        0.max(last_delete.end_index() as i64 - operation.end_index() as i64)
                            as usize,
                    );

                    Self::replace_delete_in_produced_context(
                        produced_context,
                        updated_delete.clone(),
                    );

                    updated_delete
                }
            },
        )
    }

    fn replace_delete_in_produced_context(
        produced_context: &mut MergeContext,
        delete: Option<Operation>,
    ) {
        if let Some(produced_last_delete) = produced_context.last_delete.take() {
            produced_context.shift -= produced_last_delete.len() as i64;
        }

        produced_context.last_delete = delete;
    }

    fn pick_up_dangling_delete_from_affecting_context(
        start_index: usize,
        affecting_context: &mut MergeContext,
    ) {
        match affecting_context.last_delete.as_ref() {
            Some(last_delete)
                if start_index as i64 + affecting_context.shift
                    > last_delete.end_index() as i64 =>
            {
                affecting_context.shift -= last_delete.len() as i64;
                affecting_context.last_delete = None;
            }
            _ => {}
        }
    }
}

#[cfg(test)]
mod tests {
    use std::{fs, path::Path};

    use itertools::Itertools;
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn test_calculate_operations() {
        let left = "hello world! How are you?  Adam";
        let right = "Hello, my friend! How are you doing? Albert";

        let operations = OperationSequence::from_strings(left, right);

        insta::assert_debug_snapshot!(operations);

        let mut left = Rope::from_str(left);
        let new_right = operations.apply(&mut left).unwrap();

        assert_eq!(new_right.to_string(), right);
    }

    #[test]
    fn test_calculate_operations_with_no_diff() {
        let left = "hello world!";
        let right = "hello world!";

        let operations = OperationSequence::from_strings(left, right);

        assert_eq!(operations.operations.len(), 0);

        let mut left = Rope::from_str(left);
        let new_right = operations.apply(&mut left).unwrap();

        assert_eq!(new_right.to_string(), right);
    }

    #[test]
    fn test_merges() {
        // Both replaced one token but different
        test_merge_both_ways(
            "original_1 original_2 original_3",
            "original_1 edit_1 original_3",
            "original_1 original_2 edit_2",
            "original_1 edit_1 edit_2",
        );

        // Both replaced the same one token
        test_merge_both_ways(
            "original_1 original_2 original_3",
            "original_1 edit_1 original_3",
            "original_1 edit_1 original_3",
            "original_1 edit_1 edit_1 original_3",
        );

        // One deleted a large range, the other deleted subranges and inserted as well
        test_merge_both_ways(
            "original_1 original_2 original_3 original_4 original_5",
            "original_1 original_5",
            "original_1 edit_1 original_3 edit_2 original_5",
            "original_1 edit_1 edit_2 original_5",
        );

        // One deleted a large range, the other inserted and deleted a partially overlapping range
        test_merge_both_ways(
            "original_1 original_2 original_3 original_4 original_5",
            "original_1 original_5",
            "original_1 edit_1 original_3 edit_2",
            "original_1 edit_1 edit_2",
        );

        // Merge a replace and an append
        test_merge_both_ways("a b ", "c d ", "a b c d ", "c d c d ");

        test_merge_both_ways("a b c d e", "a e", "a c e", "a e");

        test_merge_both_ways("a 0 1 2 b", "a b", "a E 1 F b", "a E F b");

        test_merge_both_ways(
            "a this one delete b",
            "a b",
            "a my one change b",
            "a my change b",
        );

        test_merge_both_ways(
            "this stays, this is one big delete, don't touch this",
            "this stays, don't touch this",
            "this stays, my one change, don't touch this",
            "this stays, my change, don't touch this",
        );

        test_merge_both_ways("1 2 3 4 5 6", "1 6", "1 2 4 ", "1 ");

        test_merge_both_ways(
            "hello world",
            "hi, world",
            "hello my friend!",
            "hi, my friend!",
        );

        // test_merge_both_ways("hello world", "world !", "hi hello world", "hi world !");

        test_merge_both_ways(
            "both delete the same word",
            "both the same word",
            "both the same word",
            "both the same word",
        );

        test_merge_both_ways("    ", "it’s utf-8!", "   ", "it’s utf-8!");

        test_merge_both_ways(
            "both delete the same word but one a bit more",
            "both the same word",
            "both same word",
            "both same wordword",
        );

        test_merge_both_ways(
            "long text with one big delete and many small",
            "long small",
            "long with big and small",
            "long small",
        );
    }

    #[test]

    fn test_merge_files_without_panic() {
        let files = vec![
            "pride_and_prejudice.txt",
            "romeo_and_juliet.txt",
            "room_with_a_view.txt",
        ];

        let root = Path::new("test/resources/");
        let contents = files
            .into_iter()
            .map(|name| fs::read_to_string(root.join(name)).unwrap())
            .map(|text| text[..15000].to_string())
            .collect::<Vec<_>>();

        contents
            .iter()
            .permutations(3)
            .unique()
            .for_each(|permutations| {
                test_merge(permutations[0], permutations[1], permutations[2]);
            });
    }

    fn test_merge_both_ways(original: &str, edit_1: &str, edit_2: &str, expected: &str) {
        assert_eq!(test_merge(original, edit_1, edit_2), expected);
        assert_eq!(test_merge(original, edit_2, edit_1), expected);
    }

    fn test_merge(original: &str, edit_1: &str, edit_2: &str) -> String {
        // println!(
        //     "original: '{:#}'",
        //     original[..100.min(original.len())].to_string()
        // );
        // println!(
        //     "edit_1: '{:#}'",
        //     edit_1[..100.min(edit_1.len())].to_string()
        // );
        // println!(
        //     "edit_2: '{:#}'",
        //     edit_2[..100.min(edit_2.len())].to_string()
        // );

        let mut original = Rope::from_str(original);

        let operations_1 = OperationSequence::from_strings(&original.to_string(), edit_1);
        // println!(
        //     "operations_1: {:#?}",
        //     operations_1.operations[..20.min(operations_1.operations.len())].to_vec()
        // );
        let operations_2 = OperationSequence::from_strings(&original.to_string(), edit_2);
        // println!(
        //     "operations_2: {:#?}",
        //     operations_2.operations[..20.min(operations_2.operations.len())].to_vec()
        // );

        assert_eq!(
            operations_1
                .apply(&mut original.clone())
                .unwrap()
                .to_string(),
            edit_1
        );
        assert_eq!(
            operations_2
                .apply(&mut original.clone())
                .unwrap()
                .to_string(),
            edit_2
        );

        let merged = operations_1.merge(&operations_2).unwrap();

        let result = merged.apply(&mut original).unwrap();
        result.to_string()
    }
}
