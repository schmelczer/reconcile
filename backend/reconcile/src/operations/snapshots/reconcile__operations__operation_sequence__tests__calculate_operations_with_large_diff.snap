---
source: reconcile/src/operations/operation_sequence.rs
expression: result
snapshot_kind: text
---
OperationSequence {
    operations: [
        (
            0,
            Insert {
                index: 0,
                text: "Hello, my friend! ",
            },
        ),
        (
            0,
            Delete {
                index: 18,
                deleted_character_count: 13,
                deleted_text: Some(
                    "hello world! ",
                ),
            },
        ),
        (
            21,
            Delete {
                index: 26,
                deleted_character_count: 5,
                deleted_text: Some(
                    "you? ",
                ),
            },
        ),
        (
            26,
            Delete {
                index: 26,
                deleted_character_count: 5,
                deleted_text: Some(
                    " Adam",
                ),
            },
        ),
        (
            31,
            Insert {
                index: 26,
                text: "you ",
            },
        ),
        (
            31,
            Insert {
                index: 30,
                text: "doing? Albert",
            },
        ),
    ],
}
